File: pkg/store/queue_store.go
package store

import (
	"context"
	"fmt"
	"time"

	"gorm.io/gorm"

	"github.com/PineappleBond/xyncra-server/pkg/store/model"
)

// QueueStore provides data access operations for the retry task queue.
type QueueStore struct {
	db *gorm.DB
}

// Save inserts a new retry task into the queue.
func (qs *QueueStore) Save(ctx context.Context, task *model.RetryTask) error {
	if err := qs.db.WithContext(ctx).Create(task).Error; err != nil {
		return classifyError(fmt.Errorf("store: save retry task: %w", err))
	}
	return nil
}

// ListPending returns retry tasks with status "pending" and NextRetry <= now,
// ordered by NextRetry ascending (soonest first).
func (qs *QueueStore) ListPending(ctx context.Context, limit int) ([]*model.RetryTask, error) {
	if limit <= 0 {
		limit = 50
	}

	var tasks []*model.RetryTask
	err := qs.db.WithContext(ctx).
		Where("status = ? AND next_retry <= ?", "pending", time.Now()).
		Order("next_retry ASC").
		Limit(limit).
		Find(&tasks).Error
	if err != nil {
		return nil, classifyError(fmt.Errorf("store: list pending retry tasks: %w", err))
	}
	return tasks, nil
}

// Update saves changes to a retry task (attempt count, next retry time,
// last error, etc.).
func (qs *QueueStore) Update(ctx context.Context, task *model.RetryTask) error {
	if err := qs.db.WithContext(ctx).Save(task).Error; err != nil {
		return classifyError(fmt.Errorf("store: update retry task: %w", err))
	}
	return nil
}

// MarkFailed sets the task's status to "failed" so it no longer appears in
// ListPending results.
func (qs *QueueStore) MarkFailed(ctx context.Context, id string, lastError string) error {
	result := qs.db.WithContext(ctx).
		Model(&model.RetryTask{}).
		Where("id = ?", id).
		Updates(map[string]any{
			"status":     "failed",
			"last_error": lastError,
		})
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: mark retry task failed: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	return nil
}

// Delete removes a retry task by its primary key.
func (qs *QueueStore) Delete(ctx context.Context, id string) error {
	result := qs.db.WithContext(ctx).Delete(&model.RetryTask{}, "id = ?", id)
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: delete retry task: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	return nil
}

// Count returns the total number of retry tasks with the given status.
func (qs *QueueStore) Count(ctx context.Context, status string) (int64, error) {
	var count int64
	err := qs.db.WithContext(ctx).
		Model(&model.RetryTask{}).
		Where("status = ?", status).
		Count(&count).Error
	if err != nil {
		return 0, classifyError(fmt.Errorf("store: count retry tasks: %w", err))
	}
	return count, nil
}

File: demo/web/packages/xyncra-client-core/src/db/queue-store.ts
/**
 * QueueStore — data access operations for the retry task queue.
 *
 * Mirrors Go QueueStore (pkg/store/queue_store.go).
 */

import { ErrNotFound } from '../errors';
import type { XyncraDatabase } from './index';
import type { RetryTask } from './models';

/**
 * QueueStore provides data access operations for the retry task queue.
 */
export class QueueStore {
  constructor(private readonly db: XyncraDatabase) {}

  /**
   * Inserts a new retry task into the queue.
   */
  async save(task: RetryTask): Promise<void> {
    await this.db.retryTasks.add(task);
  }

  /**
   * Returns retry tasks with status "pending" and next_retry <= now,
   * ordered by next_retry ascending (soonest first).
   */
  async listPending(limit: number): Promise<RetryTask[]> {
    if (limit <= 0) limit = 50;

    const now = new Date();

    const tasks = await this.db.retryTasks
      .where('status')
      .equals('pending')
      .toArray();

    // Filter by next_retry <= now.
    const due = tasks.filter((t) => new Date(t.next_retry) <= now);

    // Sort by next_retry ascending.
    due.sort(
      (a, b) =>
        new Date(a.next_retry).getTime() - new Date(b.next_retry).getTime(),
    );

    return due.slice(0, limit);
  }

  /**
   * Saves changes to a retry task (attempt count, next retry time,
   * last error, etc.).
   */
  async update(task: RetryTask): Promise<void> {
    await this.db.retryTasks.put(task);
  }

  /**
   * Sets the task's status to "failed" so it no longer appears in
   * listPending results.
   * Throws ErrNotFound if the task does not exist.
   */
  async markFailed(id: string, lastError: string): Promise<void> {
    const updated = await this.db.retryTasks
      .where('id')
      .equals(id)
      .modify((task) => {
        task.status = 'failed';
        task.last_error = lastError;
      });
    if (updated === 0) {
      throw ErrNotFound;
    }
  }

  /**
   * Removes a retry task by its primary key.
   * Throws ErrNotFound if not found.
   */
  async delete(id: string): Promise<void> {
    const existing = await this.db.retryTasks.get(id);
    if (!existing) {
      throw ErrNotFound;
    }
    await this.db.retryTasks.delete(id);
  }

  /**
   * Returns the total number of retry tasks with the given status.
   */
  async count(status: string): Promise<number> {
    return this.db.retryTasks.where('status').equals(status).count();
  }
}
